### ISSUE TITLE
Future plans of development

### ISSUE THREAD

**Issue by Raamakrishnan** [mshr-h/vscode-verilog-hdl-support#25]
We are at the verge of reaching 50,000 downloads in the marketplace. From this point, I feel it is good to have a structured development plan. The following are a few ideas that I have.

## 1. Support for more linters
* Verilator
* VCS
* NCsim

## 2. Merging System Verilog extensions
* Verilog and System Verilog are very closely related languages. The tools/linters used by Verilog can be used by System Verilog also, at the maximum by turning on a command line switch.
* Some good features like "Hover over declarations" have been added to [mshr-h/vscode-systemverilog-support](https://github.com/mshr-h/vscode-systemverilog-support) that this extension can benefit from.
* "Go To Symbol" like features can be found in [eirikpre/VSCode-SystemVerilog](https://github.com/eirikpre/VSCode-SystemVerilog)

Since there is very little work being done in bringing HDL support to VS code, it is better to pool our resources and knowledge. What do you guys say?

(Update 2018-07-07)
### System Verilog
- [x] a new branch "sv" in mshr-h/verilog
- [x] migrate syntax highlighting from mshr-h/sv
- [x] migrate snippets
- [ ] ~~migrate "Hover variable declaration"~~ (not migrating, creating new)
- [x] Parse Symbols using Ctags
- [x] Document Outline
- [x] Hover variable declaration
- [x] Definition Provider
- [x] Intellisense
- [x] migrate "Module instantiation"
- [ ] migrate "Auto alignment"
- [ ] migrate other features
- [x] Update iverilog linter to support SV
- [x] Update xvlog linter to support SV
- [x] Update modelsim linter to support SV
- [x] Update Readme, Changelog and other documentation
- [x] Merge to master and release
- [ ] Deprecate mshr-h/sv

### Linters
- [x] Verilator
- [ ] VCS
- [ ] NCsim

### Readme
- [x] Remove the list of snippets from Readme
- [x] Table of settings with description and options
- [x] Mention this issue in the Instructions for Contributors
- [x] Section for Contributors and prior work

**Comment by mshr-h** [comment]
Thank you for your great ideas. Both will make HDL support on VSCode much richer.
I think 2 is much important.
I guess bringing features in [mshr-h/vscode-systemverilog-support](https://github.com/mshr-h/vscode-systemverilog-support) into this extension is a good way.

## Step to merge these extension
1. Support syntax highlighting for `.sv` `.SV` files
2. Bring features in [mshr-h/vscode-systemverilog-support](https://github.com/mshr-h/vscode-systemverilog-support) to this extension
2. Deprecate [mshr-h/vscode-systemverilog-support](https://github.com/mshr-h/vscode-systemverilog-support)

Any ideas or comments?

**Comment by Raamakrishnan** [comment]
I think it would be wise to create a **new branch for System Verilog** and merge the new features into it. Once everything is done, it can be merged in to master and released. While new features are being ported, any other bugs/features can be added to master and released. I will create a **checklist** in my original post with all the things that needs to be done for this merging.

Maybe after we are finished with this, we would look into [eirikpre/VSCode-SystemVerilog](https://github.com/eirikpre/VSCode-SystemVerilog)

## 3. Reorganise README
* I think it is not necessary to mention all the snippets provided. 
* All the settings provided by the extension need to be tabulated properly.
* People willing to contribute should be asked to look at this issue page and plan accordingly.
* A section to mention the contributors and the resources on which this extension is based on (like previous sublime text language syntaxes, extensions etc.)

**Comment by mshr-h** [comment]
Ok.
Sounds pretty good.

**Comment by Raamakrishnan** [comment]
I have updated my original comment with a Todo list. Will start working on it soon.
We will need the help of @HansLehnert also

**Comment by HansLehnert** [comment]
Sure. Also as a future feature, it could be useful to use the "Views" feature of vscode to show the hierarchical organization of the modules.

**Comment by Raamakrishnan** [comment]
@HansLehnert [eirikpre/VSCode-SystemVerilog](https://github.com/eirikpre/VSCode-SystemVerilog) has Symbol indexing which we can use to implement this if eirikpre agrees to merge with us

**Comment by Raamakrishnan** [comment]
@mshr-h I am done with syntax highlighting and snippets. Can you create a new branch for sv so that I can give a pull request against it?

**Comment by mshr-h** [comment]
@Raamakrishnan Just created branch.
https://github.com/mshr-h/vscode-verilog-hdl-support/tree/sv

**Comment by Raamakrishnan** [comment]
I just finished porting "Hover" and started "Definition Provider" when I noticed most of the logic used is duplicated. Moreover, it just parses the document for the variable name without any regard for module or package. It works, but it is a crude implementation.

I feel like we should follow [eirikpre/VSCode-SystemVerilog](https://github.com/eirikpre/VSCode-SystemVerilog) and build a symbol tree properly so that we can implement the other features like Hover and Definition Provider easily with that.

Would you like me to get in touch with eirikpre?

**Comment by mshr-h** [comment]
Yes, thanks.

**Comment by Raamakrishnan** [comment]
Its been 2 weeks since I emailed eirikpre. There has been no response so far. Meanwhile, I started working on a symbol provider from scratch. Some more work is yet to be done, but it has been good so far. Check it out at "sv-sym" branch of my fork of this repo (https://github.com/Raamakrishnan/vscode-verilog-hdl-support/tree/sv-sym).

It needs VS Code 1.25 to work because it uses the new outline tree view

![screenshot 21](https://user-images.githubusercontent.com/9772489/43473966-2100c846-950f-11e8-9dd3-2982550ec4d3.png)
(you can see the symbols on the bottom left in the Outline View)

**Comment by Raamakrishnan** [comment]
Eirik replied saying he is not interested in joining us.
Regarding the symbol provider that I was working on, it is becoming unstable and very slow on real life verilog code. It is not possible to include all the possible ways to write something in verilog in a few regexes.
So, I feel the better way to proceed is to use the ctags file with the help of universal-ctags, parse the output and show it as symbols. Just like how the Python extension for VS code does it. What do you think about this?

**Comment by mshr-h** [comment]
@Raamakrishnan 
I understand the problem.
The solution you are thinking sounds good to me if it supports Windows.
I like UNIX Philosophy. So leveraging existing commands/tools is a good way.

**Comment by Raamakrishnan** [comment]
@mshr-h Yes it will work on Windows also. I will start working on it soon.

**Comment by aviadb** [comment]
A must have feature(s) is the capabilities of verilog-mode to Emacs.
auto-inst and auto wire connections

**Comment by terriblefire** [comment]
i just put a pull request in with verilator support.

**Comment by terriblefire** [comment]
Is it worth considering having a comment tag at the start of the file to let you choose a different verilog linter? The reason I ask is because i find verilator the best for logic files but it doesnt do a great job of testbench files. I would be inclined to prefer iverilog for that. 

Just an idea either way

**Comment by Raamakrishnan** [comment]
I understand your usage but I am not sure if many people use it this way. We will keep this in record and maybe some day we will implement it.

**Comment by Raamakrishnan** [comment]
A new extension for SystemVerilog with full support for AutoComplete, Symbols providers, definition provider and much more, all using ANTLR4 has been released recently [Rockdoor/systemverilog](https://github.com/Rockdoor/systemverilog).

I had tried using ANTLR4 before, but never got it to work with typescript. Rather, my knowledge with typescript is less. But that work looks solid. I hope we can get him to work with us.

**Comment by mshr-h** [comment]
Thanks!
Sounds good.

**Comment by Raamakrishnan** [comment]
With that last PR for linting, I think everything is done. The "Auto alignment" seems very buggy for me. I will get to that later. If everyone is fine, @mshr-h can rebase it on master. Then I can add the Readme and Changelog.

Also, now might be a good time for you guys to try out these new features!

**Comment by mshr-h** [comment]
Thanks. I'll check it this weekend.

**Comment by mshr-h** [comment]
I've just mergen 'sv' to 'master'.
Thank you all for your support!
It will be the first our major release of v1.0.0.

**Comment by realsarm** [comment]
That would be great to add some refacting feature like changing port of module and changing it else where and reanaming !

**Comment by nijawang** [comment]
If possible to add Verilog-AMS support with .vams?

**Comment by DVCoder** [comment]
Can we add custom syntax highlighting for uvm_info, uvm_error, uvm_warning, etc.?
